File: mobile/modules/device-info-helper/ios/DeviceInfoHelperModule.swift
import ExpoModulesCore
import UIKit

public class DeviceInfoHelperModule: Module {
  public func definition() -> ModuleDefinition {
    Name("DeviceInfoHelper")

    Function("getDeviceInfo") { () -> [String: String] in
      let device = UIDevice.current
      let model = device.model.trimmingCharacters(in: .whitespacesAndNewlines)
      let safeModel = model.isEmpty ? "iPhone" : model
      let osName = device.systemName.trimmingCharacters(in: .whitespacesAndNewlines)
      let safeOsName = osName.isEmpty ? "iOS" : osName
      let osVersion = device.systemVersion.trimmingCharacters(in: .whitespacesAndNewlines)
      let safeOsVersion = osVersion.isEmpty ? "unknown" : osVersion

      return [
        "platform": "ios",
        "brand": "Apple",
        "model": safeModel,
        "osName": safeOsName,
        "osVersion": safeOsVersion,
        "suggestedAlias": "\(safeModel) Owner"
      ]
    }
  }
}

File: mobile/modules/device-info-helper/src/DeviceInfoHelperModule.ts
import { requireOptionalNativeModule } from 'expo';
import { Platform } from 'react-native';

import type { DeviceInfoSnapshot } from './DeviceInfoHelper.types';

export type DeviceInfoHelperModuleType = {
  getDeviceInfo(): DeviceInfoSnapshot;
};

const fallbackModule: DeviceInfoHelperModuleType = {
  getDeviceInfo() {
    const platform = Platform.OS === 'ios' || Platform.OS === 'android' ? Platform.OS : 'web';
    const model = platform === 'ios' ? 'iPhone' : platform === 'android' ? 'Android Device' : 'Browser';

    return {
      platform,
      brand: platform === 'ios' ? 'Apple' : platform === 'android' ? 'Android' : 'Browser',
      model,
      osName: platform === 'ios' ? 'iOS' : platform === 'android' ? 'Android' : 'Web',
      osVersion: 'unknown',
      suggestedAlias: `${model} Owner`,
    };
  },
};

const nativeModule = requireOptionalNativeModule<DeviceInfoHelperModuleType>('DeviceInfoHelper');

export default nativeModule ?? fallbackModule;
